flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/CodeGenUtils.scala

```diff
@@ -0,0 +1,248 @@
+
+  def needCopyForType(t: InternalType): Boolean = t match {
+    case InternalTypes.STRING => true
+    case _: ArrayType => true
+    case _: MapType => true
+    case _: RowType => true
+    case _: GenericType[_] => true
+    case _ => false
+  }
+
+  def needCloneRefForType(t: InternalType): Boolean = t match {
+    case InternalTypes.STRING => true
+    case _ => false
+  }
+
+  // when casting we first need to unbox Primitives, for example,
+  // float a = 1.0f;
+  // byte b = (byte) a;
+  // works, but for boxed types we need this:
+  // Float a = 1.0f;
+  // Byte b = (byte)(float) a;
+  def primitiveTypeTermForType(t: InternalType): String = t match {
+    case InternalTypes.INT => "int"
+    case InternalTypes.LONG => "long"
+    case InternalTypes.SHORT => "short"
+    case InternalTypes.BYTE => "byte"
+    case InternalTypes.FLOAT => "float"
+    case InternalTypes.DOUBLE => "double"
+    case InternalTypes.BOOLEAN => "boolean"
+    case InternalTypes.CHAR => "char"
+
+    case InternalTypes.DATE => "int"
+    case InternalTypes.TIME => "int"
+    case InternalTypes.TIMESTAMP => "long"
+
+    // TODO: support INTERVAL_MONTHS and INTERVAL_MILLIS in the future
+
+    case _ => boxedTypeTermForType(t)
+  }
+
+  def boxedTypeTermForType(t: InternalType): String = t match {
+    case InternalTypes.INT => classOf[JInt].getCanonicalName
+    case InternalTypes.LONG => classOf[JLong].getCanonicalName
+    case InternalTypes.SHORT => classOf[JShort].getCanonicalName
+    case InternalTypes.BYTE => classOf[JByte].getCanonicalName
+    case InternalTypes.FLOAT => classOf[JFloat].getCanonicalName
+    case InternalTypes.DOUBLE => classOf[JDouble].getCanonicalName
+    case InternalTypes.BOOLEAN => classOf[JBoolean].getCanonicalName
+    case InternalTypes.CHAR => classOf[JChar].getCanonicalName
+
+    case InternalTypes.DATE => boxedTypeTermForType(InternalTypes.INT)
+    case InternalTypes.TIME => boxedTypeTermForType(InternalTypes.INT)
+    case InternalTypes.TIMESTAMP => boxedTypeTermForType(InternalTypes.LONG)
+
+    case InternalTypes.STRING => BINARY_STRING
+    case InternalTypes.BINARY => "byte[]"
+
+    case _: RowType => classOf[BaseRow].getCanonicalName
+    case _: DecimalType => throw new UnsupportedOperationException
+    case _: ArrayType => throw new UnsupportedOperationException
+    case _: MapType => throw new UnsupportedOperationException
+
+    case gt: GenericType[_] => gt.getTypeInfo.getTypeClass.getCanonicalName
+  }
+
+  /**
+    * Gets the boxed type term from external type info.
+    * We only use TypeInformation to store external type info.
+    */
+  def boxedTypeTermForExternalType(t: TypeInformation[_]): String = t match {
+    // From PrimitiveArrayTypeInfo we would get class "int[]", scala reflections
+    // does not seem to like this, so we manually give the correct type here.
+    case INT_PRIMITIVE_ARRAY_TYPE_INFO => "int[]"
```

Comment:
I try getTypeClass().getCanonicalName(), it will output byte[], seem it work?

Reply:
Yes, I think we can use `getTypeClass().getCanonicalName()` instead. 
Not sure why hardcode "int[]" in the original code.